central/scannerdefinitions/handler/offline.go

```diff
@@ -0,0 +1,211 @@
+package handler
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	blob "github.com/stackrox/rox/central/blob/datastore"
+	"github.com/stackrox/rox/central/scannerdefinitions/file"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+// offlineFileManager coordinates reads and writes to offline files that are backed
+// by the blob store.
+type offlineFileManager struct {
+	blobStore blob.Datastore
+
+	// dataDir the base directory where offline files will be written in prep for reads.
+	dataDir string
+
+	// files maps a blob's name to a file.
+	files map[string]*offlineFile
+}
+
+// offlineFileOpenerFunc returns a handle to a freshly opened file and its size.
+type offlineFileOpenerFunc func() (io.ReadCloser, int64, error)
+
+// newOfflineFileManager creates a new offline file manager.
+func newOfflineFileManager(blobStore blob.Datastore, dataDir string) *offlineFileManager {
+	return &offlineFileManager{
+		files:     map[string]*offlineFile{},
+		blobStore: blobStore,
+		dataDir:   dataDir,
+	}
+}
+
+// Register registers a blob name to be managed. A blob name must be registered before
+// the associated offline file can be used.
+func (o *offlineFileManager) Register(blobName string) {
+	o.files[blobName] = newOfflineFile(o.dataDir, o.blobStore, blobName)
```

Comment:
interested in adding any protections to ensure only unique blob names are used (like adding a panic)?